File: jun1/src/bitset256.rs
use std::ops::{BitAnd, BitAndAssign, BitOr, BitOrAssign, BitXor, Not};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BitSet256 {
    x: u128,
    y: u128,
}

impl BitSet256 {
    pub fn is_set(&self, index: u8) -> bool {
        if index < 128 {
            self.x & (1 << index) != 0
        } else {
            self.y & (1 << (index - 128)) != 0
        }
    }

    pub fn set_bit(&mut self, index: u8) {
        if index < 128 {
            self.x |= 1 << index;
        } else {
            self.y |= 1 << (index - 128);
        }
    }

    pub fn clear_bit(&mut self, index: u8) {
        if index < 128 {
            self.x &= !(1 << index);
        } else {
            self.y &= !(1 << (index - 128));
        }
    }

    pub fn update(&mut self, other: &Self) {
        self.x |= other.x;
        self.y |= other.y;
    }

    pub fn len(&self) -> usize {
        self.x.count_ones() as usize + self.y.count_ones() as usize
    }

    pub fn is_empty(&self) -> bool {
        self.x == 0 && self.y == 0
    }

    pub fn clear(&mut self) {
        self.x = 0;
        self.y = 0;
    }

    pub fn all() -> Self {
        BitSet256 { x: u128::MAX, y: u128::MAX }
    }

    pub fn none() -> Self {
        BitSet256 { x: 0, y: 0 }
    }
}

impl BitOr for BitSet256 {
    type Output = Self;

    fn bitor(self, other: Self) -> Self {
        BitSet256 {
            x: self.x | other.x,
            y: self.y | other.y,
        }
    }
}

impl BitAnd for BitSet256 {
    type Output = Self;

    fn bitand(self, other: Self) -> Self {
        BitSet256 {
            x: self.x & other.x,
            y: self.y & other.y,
        }
    }
}

impl BitXor for BitSet256 {
    type Output = Self;

    fn bitxor(self, other: Self) -> Self {
        BitSet256 {
            x: self.x ^ other.x,
            y: self.y ^ other.y,
        }
    }
}

impl Not for BitSet256 {
    type Output = Self;

    fn not(self) -> Self {
        BitSet256 {
            x: !self.x,
            y: !self.y,
        }
    }
}

impl BitAndAssign for BitSet256 {
    fn bitand_assign(&mut self, other: Self) {
        self.x &= other.x;
        self.y &= other.y;
    }
}

impl BitOrAssign for BitSet256 {
    fn bitor_assign(&mut self, other: Self) {
        self.x |= other.x;
        self.y |= other.y;
    }
}

File: jun1/src/bitset256.py
from typing import Optional


class BitSet256:
    def __init__(self, x: Optional[int] = None, y: Optional[int] = None) -> None:
        self.x = x or 0
        self.y = y or 0

    def is_set(self, index: int) -> bool:
        if index < 128:
            return self.x & (1 << index) != 0
        return self.y & (1 << (index - 128)) != 0

    def set_bit(self, index: int) -> None:
        if index < 128:
            self.x |= 1 << index
        else:
            self.y |= 1 << (index - 128)

    def clear_bit(self, index: int) -> None:
        if index < 128:
            self.x &= ~(1 << index)
        else:
            self.y &= ~(1 << (index - 128))

    def update(self, other: 'BitSet256') -> None:
        self.x |= other.x
        self.y |= other.y

    def __len__(self) -> int:
        return self.x.bit_count() + self.y.bit_count()

    def is_empty(self) -> bool:
        return self.x == 0 and self.y == 0

    def clear(self) -> None:
        self.x = 0
        self.y = 0

    @staticmethod
    def all() -> 'BitSet256':
        return BitSet256(x=(1 << 128) - 1, y=(1 << 128) - 1)

    @staticmethod
    def none() -> 'BitSet256':
        return BitSet256()

    def __or__(self, other: 'BitSet256') -> 'BitSet256':
        return BitSet256(self.x | other.x, self.y | other.y)

    def __and__(self, other: 'BitSet256') -> 'BitSet256':
        return BitSet256(self.x & other.x, self.y & other.y)

    def __invert__(self) -> 'BitSet256':
        return BitSet256(~self.x, ~self.y)

    def __ior__(self, other: 'BitSet256') -> 'BitSet256':
        self.x |= other.x
        self.y |= other.y
        return self

    def __iand__(self, other: 'BitSet256') -> 'BitSet256':
        self.x &= other.x
        self.y &= other.y
        return self
